pubsub/mqtt/mqtt.go

```diff
@@ -0,0 +1,168 @@
+
+package mqtt
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/dapr/pkg/logger"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+)
+
+const (
+	// Keys
+	mqttURL          = "url"
+	mqttQOS          = "qos"
+	mqttRetain       = "retain"
+	mqttClientID     = "consumerID"
+	mqttCleanSession = "cleanSession"
+
+	// errors
+	errorMsgPrefix = "MQTT pub sub error:"
+
+	// Defaults
+	defaultQOS          = 0
+	defaultRetain       = false
+	defaultWait         = 3 * time.Second
+	defaultCleanSession = true
+)
+
+// mqttPubSub type allows sending and receiving data to/from MQTT broker.
+type mqttPubSub struct {
+	client   mqtt.Client
+	metadata *metadata
+	logger   logger.Logger
+}
+
+// NewMQTTPubSub returns a new mqttPubSub instance.
+func NewMQTTPubSub(logger logger.Logger) pubsub.PubSub {
+	return &mqttPubSub{logger: logger}
+}
+
+func parseMQTTMetaData(md pubsub.Metadata) (*metadata, error) {
+	m := metadata{}
+
+	// required configuration settings
+	if val, ok := md.Properties[mqttURL]; ok && val != "" {
+		m.url = val
+	} else {
+		return &m, fmt.Errorf("%s missing url", errorMsgPrefix)
+	}
+
+	// optional configuration settings
+	m.qos = defaultQOS
+	if val, ok := md.Properties[mqttQOS]; ok && val != "" {
+		qosInt, err := strconv.Atoi(val)
+		if err != nil {
+			return &m, fmt.Errorf("%s invalid qos %s, %s", errorMsgPrefix, val, err)
+		}
+		m.qos = byte(qosInt)
+	}
+
+	m.retain = defaultRetain
+	if val, ok := md.Properties[mqttRetain]; ok && val != "" {
+		var err error
+		m.retain, err = strconv.ParseBool(val)
+		if err != nil {
+			return &m, fmt.Errorf("%s invalid retain %s, %s", errorMsgPrefix, val, err)
+		}
+	}
+
+	m.clientID = uuid.New().String()
+	if val, ok := md.Properties[mqttClientID]; ok && val != "" {
+		m.clientID = val
+	}
+
+	m.cleanSession = defaultCleanSession
+	if val, ok := md.Properties[mqttCleanSession]; ok && val != "" {
+		var err error
+		m.cleanSession, err = strconv.ParseBool(val)
+		if err != nil {
+			return &m, fmt.Errorf("%s invalid clean session %s, %s", errorMsgPrefix, val, err)
+		}
+	}
+
+	return &m, nil
+}
+
+// Init parses metadata and creates a new Pub Sub client.
+func (m *mqttPubSub) Init(metadata pubsub.Metadata) error {
+	mqttMeta, err := parseMQTTMetaData(metadata)
+	if err != nil {
+		return err
+	}
+	m.metadata = mqttMeta
+
+	uri, err := url.Parse(m.metadata.url)
+	if err != nil {
+		return err
+	}
+	client, err := m.connect(uri)
+	if err != nil {
+		return err
+	}
+
+	m.client = client
+
+	m.logger.Debug("MQTT message bus initialization complete")
+	return nil
+}
+
+// Publish the topic to mqtt pub sub.
+func (m *mqttPubSub) Publish(req *pubsub.PublishRequest) error {
+	m.logger.Debugf("mqtt: publishing topic %v with data: %v", req.Topic, req.Data)
+
+	token := m.client.Publish(req.Topic, m.metadata.qos, m.metadata.retain, req.Data)
+	for !token.WaitTimeout(defaultWait) {
```

Comment:
I'm curious, what's the reason for calling this in a loop instead of just `token.Wait()` instead?

Reply:
It is the same.
Though we can modify the logic to look like:
```go
if !token.WaitTimeout(defaultWait) && token.Error() != nil {
		return nil, token.Error()
}
```
What do you suggest ?

Reply:
I support the modification. @LMWF wdyt?

Reply:
Agree.

Reply:
Done. Thanks